### ISSUE TITLE
Multi thread Chunker for large single files?

### ISSUE THREAD

**Issue by BizClimb** [rclone/rclone#5041]
#### What is your current rclone version (output from `rclone version`)?
1.54

#### What problem are you are trying to solve?
Since various cloud storage providers limit upload speed per upload stream (not total for the account), you can speed up large file transfers by multi threading chunks instead of going one by one.

#### How do you think rclone should be changed to solve that?
Multi thread chunker to upload multiple parts at once instead of part by part 1 by 1 for large files.

**Comment by darthShadow** [comment]
Anything different from https://github.com/rclone/rclone/issues/4798?

**Comment by ivandeex** [comment]
Pending on MultithreadPut interface proposed by #4798

Note: MD5/SHA1 hashsums are sequential by nature and cannot be supported by multithread uploads.

**Comment by ivandeex** [comment]
As noted in https://github.com/rclone/rclone/issues/4798#issuecomment-733641543, this feature is "orthogonal" to the upload resume interface.

**Comment by ivandeex** [comment]
A sketch of multi-thread implementation for chunker can base on ideas similar to https://github.com/rclone/rclone/pull/4547#issuecomment-786593956

We add a new _per-upload_ (in chunker code it will be _per-transaction_) control chunk to save info about:
- sizes and offsets of uploaded chunks
- hashes of uploaded chunks taken from the underlying backend (since chunks are files from its perspective

This can be used to implement a process sketched in https://github.com/rclone/rclone/issues/4798#issuecomment-733641543 or even its extension (multi-part resume) proposed in https://forum.rclone.org/t/intelligent-faster-chunker-file-updates-on-checksum-enabled-remotes/22313

The latter proposal, besides the multi-part resume case, can be useful in the use-case of uploading an append-only file. However, it won't work in general, when the file to re-upload is modified at random. That case will need a technique known as Content-Defined Chunking (CDC), which is out of scope for this ticket.

**Comment by ivandeex** [comment]
A comment on hashing.

Unlike the _Resumer_ case explained in https://github.com/rclone/rclone/pull/4547#issuecomment-786889739, we cannot use a partial rolling hash for the upload as a whole as traditional hashes do not support multi-scatter mode.

As noted in https://forum.rclone.org/t/intelligent-faster-chunker-file-updates-on-checksum-enabled-remotes/22313/7, most cloud storage providers are tied to traditional hashsums and simply stop checksumming when it comes to multipart uploads.

As explained by https://github.com/rclone/rclone/pull/4656#issuecomment-769631364, dropbox in their recent API solves this problem introducing an _array of hashes_. This is a step forward. Chunker could provide a similar API and return array with hashes of chunk files from the underlying backend.

I wrote in https://github.com/rclone/rclone/pull/4656#issuecomment-769921726 that we can take a step in different direction and introduce [Blake3](https://github.com/BLAKE3-team/BLAKE3) support in chunker as this hash is internally [constructed](https://github.com/oconnor663/bao/blob/master/docs/spec_0.9.1.md#tree-structure) as a kind [merkle tree](https://en.wikipedia.org/wiki/Merkle_tree) and consequently supports multi-threading.

Blake3 has a number of advantages in comparison with hash arrays:
- it can be verified by a publicly available [b3sum](https://aur.archlinux.org/packages/b3sum/) uitility available in some distributions (and we can implement it in `rclone hash` for others where not),
- it can be recorded in and checked with a `SUM` file
- it's relatively small (256 bits) and highly performant.

However, enabling it for multi-part uploads will involve a number of tricky decisions:
- develop a format to capture levels of a partial merkle tree in a dedicated control chunk
- develop a technique to split partial merkle tree between already and not-yet uploaded chunks
- embed state-capturing hooks in a blake3 golang package, probably forking is the only solution

**Comment by Zegorax** [comment]
Out of curiosity, is there any ETA for this issue and/or the blocker ?

**Comment by emailkanji** [comment]
@Zegorax 
> Out of curiosity, is there any ETA for this issue and/or the blocker ?

Keep an eye on the two following issues. #87 and #4798. I think the former is nearing basic implementation. After that the later will probably need to be completed before this issue is tackled. With the former being a WIP since 2016 don't expect a short timeline.

As a work around you use an rclone union, chunker, move, and a local directory. You'll upload to chunker which writes chunks in the union to the local directory then actual uploading will use rclone move to multi-thread the upload of the underlying chunked files to the remotes; without the union nor chunker. And chunker shouldn't even know they were moved.

So it'll be remote: then union: of [remote:nc and a local directory] and then chunker on top of the union: along with a move triggered somehow.

You might want the union cache time to be pretty short so chunker doesn't try to read from the local directory but I don't know how union handles cached files being missing.

It is extra janky but might get the job done for now.

**Comment by eharris** [comment]
Since chunker splits large files into separate parts, and past the chunker layer each of those smaller files ARE actually a whole single file (as far as the endpoint is concerned), it would seem that at least those chunked parts could be uploaded individually, is that not the case?  Why can't this be done without needing to solve the larger and more complex problem of being able to multi-thread upload parts of a single file?

**Comment by ncw** [comment]
@eharris I think actually implementing the `OpenChunkWriter` interface might be easier and this would allow proper parallel uploads.

https://github.com/rclone/rclone/blob/8817ee25ae245b15497bc08e48f57acca5d8530a/fs/features.go#L171-L176

Do you want to have a go at this?

**Comment by Rootax** [comment]
Hello,

Any news on this one ? I'm using chuncker to help the balancing on a union wrapped drive, and yeah the single upload at a time is really not speed friendly. Not a critic, it's just to say that I believe a lot of ppl are still interested in this feature request :)

**Comment by TTTPOB** [comment]
Edit: I found that I do not have enough time to finish this properly, so there will not be a PR from me.

My original problem was that any backend-side failure could trigger a full file re-upload, which made all successfully uploaded chunks useless. I drafted what I think is a cleaner design below, but finishing it would take more time than I can spend on it right now. I am a PhD student after all.

In the end, I used an ugly local workaround for my own case. I added a `--chunker-low-level-retries` CLI argument, which solves the practical problem I had, even though it is not the design I would want to upstream.

Since coding agents are much more common today, I think anyone who reads this issue and wants the feature can probably pick up the idea and implement a proper version. I just wanted to leave this note here so it is clear that I am no longer working on it.


I'm working on this and plan to open a PR soon.

The main issue I have run into is hash handling for `hash_type = md5` and
`hash_type = sha1`.

Chunker metadata stores the hash of the reconstructed composite object, not
per-chunk hashes. For MD5 and SHA1, the digest has to be updated in file order.
However, once chunker is driven through the multi-threaded copy / `ChunkWriter`
path, chunks may be written out of order.

That means the writer cannot simply update the composite MD5/SHA1 digest inside
each `WriteChunk` call unless it also preserves the original byte order. The
possible approaches I see are:

1. use the source object's hash when it is available;
2. fall back to single-threaded writing when chunker has to compute MD5/SHA1
   itself;
3. buffer/reorder chunk data, which can become a large RAM or disk cache;
4. reopen source ranges and feed them to the hasher in order, which adds extra
   reads and depends on range-read support.

When there is no src fast hash available, and user forced multi-threaded copy, i plan to force the user to choose which way to solve the hash problem.

Even furthur, I am planning to introduce blake3 hash option to chunker, so that it natively support multi-threaded copy.
